README.md
# ffxivautocraft
Automate long, tedious crafts and get real-time progress updates in the terminal.

## Usage
```
usage: autocraft.exe [-h] [-click_delay CLICK_DELAY] [-craft_delay CRAFT_DELAY] [-initial_delay INITIAL_DELAY] craft_count craft_length macro_key

Automate basic crafting task with macro. For crafts spanning multiple macros, 
I recommend using the /nextmacro command provided by the Macro Chain Dalamud plugin.

positional arguments:
  craft_count           
        number of items to craft
  craft_length
        the duration of the craft macro
  macro_key             
        the key corresponding to the crafting macro

optional arguments:
  -h            
        show this help message and exit
  --click_delay CLICK_DELAY
        the time (in seconds) to wait after automated clicking
  --craft_delay CRAFT_DELAY
        the time (in seconds) to wait after finishing an automated crafting macro
  --initial_delay INITIAL_DELAY
        the time (in seconds) before starting the craft automation
```

## Example
Let's say you have your crafting macro (or starting macro for chained macros) placed on your hotbars bound to the 5 key, you are crafting 30 of the item, and you know each craft takes 20 seconds. To use autocraft for this situation, follow the steps below:

1. Launch FFXIV if you haven't already done so. 
2. Position yourself somewhere private, like an inn or FC room.
3. Open your crafting log and select the item you wish to craft.
5. Open a terminal (ideally on a second monitor where you can see the progress echoed in real-time) in the directory where you've placed autocraft.exe, and run the following command:
   `autocraft.exe 30 20 5`.
6. Alt+Tab back to FFXIV and hover your mouse over the center region of the `Synthesize` button.
7. Profit.

## Notes
- The easiest ways to determine the craft length are to sum the numbers in the <wait.x> commands of the macro(s), time it with a stopwatch or make the macro in teamcraft and use the reported craft duration.
- The crafts may be unable to continue if you run out of your starting material quality (HQ or NQ) and have to change the material quality in use for the craft.


CHANGELOG.md
# Release Notes
- Added Github Actions workflows for automated release creation, building, and deployment


ffxivautocraft.go
package main

import (
	"flag"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/go-vgo/robotgo"
)

/*
Display custom usage information for the command-line interface.
*/
func usage() {
	usageCommandString := "Usage: %s [-h] [-click_delay CLICK_DELAY] " +
		"[-craft_delay CRAFT_DELAY] [-initial_delay INITIAL_DELAY] " +
		"craft_count craft_length macro_key"
	fmt.Printf(usageCommandString+"\n\n", os.Args[0])

	programDescription := "Automate basic crafting task with macro. " +
		"For crafts spanning multiple macros,\n" +
		"I recommend using the /nextmacro command provided by the " +
		"Macro Chain Dalamud plugin."
	fmt.Println(programDescription + "\n")

	fmt.Println("positional arguments:")
	fmt.Println("  craft_count int\n\tthe number of items to craft")
	fmt.Println("  craft_length float\n\tthe duration of the crafting macro")
	fmt.Println("  macro_key string\n\tthe key corresponding to the crafting macro")

	fmt.Println("\noptional arguments:")
	flag.PrintDefaults()
}

func main() {
	// initialize constants
	const (
		PREFIX            = "Progress: "
		DECIMAL_PRECISION = 0
		BAR_LENGTH        = 30
		BAR_CHAR          = "█"
	)

	//
	// setup command-line interface
	//

	// define positional arguments
	var craftCount int
	var craftLength float64
	var macroKey string

	// define optional arguments
	var clickDelay float64
	flag.Float64Var(&clickDelay, "click_delay", 1.0,
		"the time (in seconds) to wait after automated clicking")
	var craftDelay float64
	flag.Float64Var(&craftDelay, "craft_delay", 4.0,
		"the time (in seconds) to ait after finishing an automated crafting macro")
	var initialDelay float64
	flag.Float64Var(&initialDelay, "initial_delay", 10.0,
		"the time (in seconds) before starting the craft automation")

	// enable custom usage printout
	flag.Usage = usage

	// parse args and store positional arguments in their corresponding variables
	flag.Parse()

	craftCount, _ = strconv.Atoi(flag.Arg(0))
	craftLength, _ = strconv.ParseFloat(flag.Arg(1), 64)
	macroKey = flag.Arg(2)

	// setup random number generation
	rand.Seed(time.Now().UnixNano())

	// wait for initial delay
	robotgo.MilliSleep(int(initialDelay * 1000.0))

	// start automated crafting loop
	itemsCrafted := 0
	for itemsCrafted < craftCount {
		// estimate and display remaining craft time
		estimatedTimeRemaining := estimateTime(craftCount, itemsCrafted,
			clickDelay, craftDelay, craftLength)
		fmt.Printf("Estimated time left for craft completion: %dm%ds.",
			estimatedTimeRemaining[0], estimatedTimeRemaining[1])

		// click the button to start crafting
		robotgo.Click()
		robotgo.MilliSleep(int(clickDelay*1000.0 + rand.Float64()*400))

		// start the crafting macro
		robotgo.KeyTap(macroKey)
		robotgo.MilliSleep(int(500.0 + rand.Float64()*200.0))

		// wait until the crafting macro is finished
		robotgo.MilliSleep(int(craftLength*1000.0 + rand.Float64()*2000.0))

		// increment crafted item count and display progress bar
		itemsCrafted++
		SUFFIX := fmt.Sprintf(" | %d/%d Items Crafted", itemsCrafted, craftCount)
		progressBar := createProgressBar(itemsCrafted, craftCount, PREFIX,
			SUFFIX, DECIMAL_PRECISION, BAR_LENGTH, BAR_CHAR)
		fmt.Println(progressBar)

		// wait before starting crafting for next item
		robotgo.MilliSleep(int(craftDelay*1000.0 + rand.Float64()*400))
	}

	fmt.Println("Crafting complete!")

	os.Exit(0)

}

/*
Create a string representing a progress bar given the current progress, the
total progress goal, and other properties of the bar.

A representation of a created bar is as follows:
Prefix |****      | 40.0% Suffix
*/
func createProgressBar(progress int, total int, prefix string, suffix string,
	decimalPrecision int, barLength int, barChar string) string {

	percent := float64(progress) / float64(total)
	barCount := int(percent * float64(barLength))
	barsStr := strings.Repeat(barChar, barCount)
	spacesStr := strings.Repeat(" ", barLength-barCount)
	formattedPercent := strconv.FormatFloat(percent * 100.0, 'f', decimalPrecision, 64)

	progressBarStr := fmt.Sprintf("%s|%s%s| %s%%%s", prefix, barsStr, spacesStr, formattedPercent, suffix)

	return progressBarStr
}

/*
Estimate the time remaining to complete all crafts in minutes and seconds.
*/
func estimateTime(craftCount int, craftProgress int, clickDelay float64,
	craftDelay float64, craftLength float64) [2]int {

	timePerCraft := (0.001 + 0.05) + (0.001 + 1.2*clickDelay) +
		(0.001 + 0.1) + (craftLength + 1.0) + (1.4 * craftDelay)

	craftsRemaining := craftCount - craftProgress

	timeRemaining := float64(craftsRemaining) * timePerCraft

	minutes := int(timeRemaining / 60.0)
	seconds := int(timeRemaining - float64(minutes)*60.0)

	return [2]int{minutes, seconds}
}


autocraft/autocraft.py
import argparse as ap
import numpy as np
from pynput.keyboard import Controller as KBController
from pynput.mouse import Button as MButton, Controller as MController
import random
import time
from typing import Tuple


def progress_bar(progress: int, total: int, prefix: str = "", suffix: str = "", 
                 decimals: int = 1, bar_length: int = 30, bar_char: str = "█") -> None:
    """Create and print a progress bar.

    Args:
        progress (int): the progress toward the goal
        total (int): the goal
        prefix (str): the string to be printed before the progress bar
        suffix (str): the string to be printed after the progress bar
        decimals (int): the number of decimal places to display in the 
            percentage progress display
        bar_length (int): the horizontal width of the progress bar
        bar_char (str): the character to use for units on the progress bar
    """

    percent = progress / total
    bar_count = int(percent * bar_length)
    bars = bar_char * bar_count
    spaces = " " * (bar_length - bar_count)

    print(f"{prefix}|{bars}{spaces}| {percent:.{decimals}%}{suffix}")

    return None

def estimate_time(craft_count: int, craft_progress: int, click_delay: float, 
                  craft_delay: float, craft_length: float) -> Tuple[int, int]:
    """Estimate the time remaining in minutes and seconds.

    Args:
        craft_count (int): the number of crafts for the session
        craft_progress (int): the number of crafts completed
        click_delay (float): the delay after clicking
        craft_delay (float): the delay after finishing a craft

    Returns:
        A tuple of the form (minutes, seconds) representing the remaining 
            time to complete scheduled crafts.
    """

    time_per_craft = (
        (0.001 + 0.05) + (0.001 + 1.2 * click_delay) + (0.001 + 0.1) 
         + (craft_length + 1.0) + (1.4 * craft_delay))
    
    crafts_remaining = craft_count - craft_progress

    time_remaining = crafts_remaining * time_per_craft

    minutes = int(time_remaining / 60.0)
    seconds = int(time_remaining - minutes * 60.0)

    return (minutes, seconds)


if __name__ == "__main__":

    # setup command-line argument parsing
    parser = ap.ArgumentParser(
        description="Automate basic crafting task with macro. For crafts "
                    "spanning multiple macros, I recommend using the "
                    "/nextmacro command provided by the Macro Chain Dalamud "
                    "plugin.")
    parser.add_argument("craft_count", type=int, 
                        help="number of items to craft")
    parser.add_argument("craft_length", type=float, 
                        help="the duration of the craft macro")
    parser.add_argument("macro_key", type=str, 
                        help="the key corresponding to the crafting macro")
    parser.add_argument("--click_delay", "-clkdel", type=float, default=1.0,
                        help="the time (in seconds) to wait after automated clicking")
    parser.add_argument("--craft_delay", "-crftdel", type=float, default=4.0,
                        help="the time (in seconds) to wait after finishing an automated crafting macro")
    parser.add_argument("--initial_delay", "-initdel", type=float, default=10.0,
                        help="the time (in seconds) before starting the craft automation")
    args = parser.parse_args()
    
    # setup random number generation
    rng = np.random.default_rng(random.randint(0,100000000))

    # setup input management
    mouse = MController()
    kb = KBController()

    # wait for initial delay
    time.sleep(args.initial_delay)

    # start automated crafting loop
    items_crafted = 0
    while items_crafted < args.craft_count:
        # estimate and display remaining craft time
        estimated_time_remaining = estimate_time(args.craft_count, items_crafted, args.click_delay, args.craft_delay, args.craft_length)
        print(f"Estimated time left for craft completion: {estimated_time_remaining[0]}m{estimated_time_remaining[1]}s.")

        # click the button to start crafting
        mouse.press(MButton.left)
        time.sleep(0.001 + rng.random() * 0.1)
        mouse.release(MButton.left)

        # wait before starting the crafting macro
        time.sleep(args.click_delay + rng.random() * 0.4 * args.click_delay)

        # start the crafting macro
        kb.press(args.macro_key)
        time.sleep(0.001 + rng.random() * 0.2)
        kb.release(args.macro_key)

        # wait until the crafting macro is finished
        time.sleep(args.craft_length + rng.random() * 2.0)

        # increment crafted item count and display progress bar
        items_crafted += 1
        progress_bar(
            items_crafted, args.craft_count, prefix="Progress: ", 
            suffix=f" {items_crafted}/{args.craft_count} Items Crafted")

        # wait before starting crafting for next item
        time.sleep(args.craft_delay + rng.random() * 0.4 * args.craft_delay)
    
    print("Crafting complete!")

    # wait for the user to press a key to quit.
    input("Press any key to quit.")
